metamask-predicate/tests/harness.rs

```diff
@@ -0,0 +1,204 @@
+
+    let output_coin = Output::coin(predicate_address, amount_to_predicate, asset_id);
+    let output_change = Output::change(wallet.address().into(), 0, asset_id.into());
+
+    let mut tx = ScriptTransactionBuilder::prepare_transfer(
+        wallet_coins,
+        vec![output_coin, output_change],
+        TxParameters::default()
+            .set_gas_price(1)
+            .set_gas_limit(1_000_000),
+    )
+    .set_script(op::ret(RegId::ONE).to_bytes().to_vec())
+    .build()
+    .unwrap();
+
+    wallet.sign_transaction(&mut tx).unwrap();
+    wallet
+        .provider()
+        .unwrap()
+        .send_transaction(&tx)
+        .await
+        .unwrap();
+}
+
+async fn submit_to_predicate(
+    predicate_code: Vec<u8>,
+    predicate_address: Address,
+    wallet: &Wallet,
+    amount_to_predicate: u64,
+    asset_id: AssetId,
+    receiver_address: Address,
+    predicate_data: UnresolvedBytes,
+) {
+    let filter = ResourceFilter {
+        from: predicate_address.into(),
+        asset_id,
+        amount: amount_to_predicate,
+        ..Default::default()
+    };
+
+    let utxo_predicate_hash = wallet
+        .provider()
+        .unwrap()
+        .get_spendable_resources(filter)
+        .await
+        .unwrap();
+
+    let mut inputs = vec![];
+    let mut total_amount_in_predicate = 0;
+
+    for resource in utxo_predicate_hash {
+        match &resource {
+            Resource::Coin(coin) => {
+                inputs.push(Input::resource_predicate(
+                    resource.clone(),
+                    predicate_code.to_vec(),
+                    predicate_data.clone(),
+                ));
+                total_amount_in_predicate += coin.amount;
+            }
+            Resource::Message(_) => {}
+        }
+    }
+
+    let output_coin = Output::coin(receiver_address, total_amount_in_predicate, asset_id);
+    let output_change = Output::change(predicate_address, 0, asset_id);
+
+    let params = wallet
+        .provider()
+        .unwrap()
+        .consensus_parameters()
+        .await
+        .unwrap();
+    let new_tx = ScriptTransactionBuilder::prepare_transfer(
+        inputs,
+        vec![output_coin, output_change],
+        TxParameters::default().set_gas_limit(1_000_000),
+    )
+    .set_consensus_parameters(params)
+    .build()
+    .unwrap();
+
+    let _call_result = wallet.provider().unwrap().send_transaction(&new_tx).await;
+}
+
+async fn get_balance(wallet: &Wallet, address: Address, asset_id: AssetId) -> u64 {
+    wallet
+        .provider()
+        .unwrap()
+        .get_asset_balance(&address.into(), asset_id)
+        .await
+        .unwrap()
+}
+
+#[tokio::test]
+async fn valid_predicate_data_simple() {
+    // let msg_hash: Bits256 = Bits256::from_hex_str("0x9c411706d71aea01a27dab34a5938dbe39de195ec6f8e077a99e89f24ddef48b").unwrap();
+
+    let msg_hash: Bits256 = Bits256::from_hex_str("0xee45573606c96c98ba970ff7cf9511f1b8b25e6bcd52ced30b89df1e4a9c4323").unwrap();
```

Comment:
These seemingly random addresses should be factored out into an array and indexed. It's difficult to tell what they are and why they are different.